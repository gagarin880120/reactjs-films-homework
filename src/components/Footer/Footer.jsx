import React, { useRef, useEffect } from 'react';
import PropTypes from 'prop-types';
import styles from './Footer.module.scss';

export default function Footer({
  results, onIntersect, currentPage, query, areMoviesLoaded, url, totalPages,
}) {
  const footer = useRef();
  useEffect(() => {
    if (currentPage === totalPages) {
      return;
    }
    let observer = new IntersectionObserver((entries) => {
      entries.forEach((entry) => {
        const { isIntersecting } = entry;
        if (isIntersecting) {
          onIntersect(url, currentPage + 1, query, results);
          observer = observer.disconnect();
        }
      });
    });
    observer.observe(footer.current);
  }, [areMoviesLoaded]);

  return (
    <div className={styles.wrapper} ref={footer}>

      <div className={styles.header}>
        films
      </div>
      <p className={styles.text}>
        Copyright Â© 2020 FILMS. Andrei Mandryk
      </p>
    </div>
  );
}

Footer.propTypes = {
  results: PropTypes.instanceOf(Array),
  currentPage: PropTypes.number,
  onIntersect: PropTypes.func,
  query: PropTypes.string,
  areMoviesLoaded: PropTypes.bool,
  url: PropTypes.string,
  totalPages: PropTypes.number,
};

Footer.defaultProps = {
  results: [],
  currentPage: 1,
  onIntersect: null,
  query: '',
  areMoviesLoaded: false,
  url: '',
  totalPages: 0,
};
