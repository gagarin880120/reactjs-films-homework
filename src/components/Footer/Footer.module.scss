.wrapper {
  background: #fff;
  text-align: center;
  padding: 2%;
}

.header {
  text-transform: uppercase;
  color: #47c3f0;
  font-size: 32px;
}

.text {
  padding: 1%;
  font-size: 18px;
  color: #A7AAAB;
}
