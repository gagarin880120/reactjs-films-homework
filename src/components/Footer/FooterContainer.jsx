import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import Footer from './Footer';
import { getMovies, getMoviesBySearch, getMoviesByGenre } from '../../redux/actions';
import {
  resultsSelector, currentPageSelector, currentAPIRequestSelector,
  areMoviesLoadedSelector, totalPagesSelector,
} from '../../redux/selectors';

export function FooterContainer({
  results, onIntersect, currentPage,
  areMoviesLoaded, totalPages, currentAPIRequest,
}) {
  return (
    <Footer
      results={results}
      currentPage={currentPage}
      onIntersect={onIntersect}
      areMoviesLoaded={areMoviesLoaded}
      totalPages={totalPages}
      currentAPIRequest={currentAPIRequest}
    />
  );
}

export const mapStateToProps = (state) => ({
  results: resultsSelector(state),
  currentPage: currentPageSelector(state),
  areMoviesLoaded: areMoviesLoadedSelector(state),
  totalPages: totalPagesSelector(state),
  currentAPIRequest: currentAPIRequestSelector(state),
});

export const mapDispatchToProps = (dispatch) => ({
  onIntersect(currentAPIRequest, page, results) {
    if (currentAPIRequest.includes('search')) {
      dispatch(getMoviesBySearch(currentAPIRequest, page, results));
    } else if (currentAPIRequest.includes('genre')) {
      dispatch(getMoviesByGenre(currentAPIRequest, page, results));
    } else {
      dispatch(getMovies(currentAPIRequest, page, results));
    }
  },
});

FooterContainer.propTypes = {
  results: PropTypes.instanceOf(Array),
  currentPage: PropTypes.number,
  onIntersect: PropTypes.func,
  areMoviesLoaded: PropTypes.bool,
  totalPages: PropTypes.number,
  currentAPIRequest: PropTypes.string,
};

FooterContainer.defaultProps = {
  results: [],
  currentPage: 1,
  onIntersect: null,
  areMoviesLoaded: false,
  totalPages: 0,
  currentAPIRequest: '',
};

export default connect(mapStateToProps, mapDispatchToProps)(FooterContainer);
