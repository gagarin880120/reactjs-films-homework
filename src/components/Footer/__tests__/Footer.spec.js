import ShallowRenderer from 'react-test-renderer/shallow';
import TestRenderer, { act } from 'react-test-renderer';
import React from 'react';
import Footer from '../Footer';

describe('Footer component', () => {
  test('should render without crashing', () => {
    const renderer = new ShallowRenderer();
    renderer.render(<Footer />);
    const result = renderer.getRenderOutput();
    expect(result).toMatchSnapshot();
  });
  describe('testRenderer', () => {
    const testRenderer = TestRenderer.create(
    <Footer
      areMoviesLoaded={false}
      currentPage={5}
      totalPages={5}
    />
    );

    test('when areMoviesLoaded is updated', () => {
      act(() => {
        testRenderer.update(
          <Footer
            areMoviesLoaded={true}
            currentPage={5}
            totalPages={5}
          />
        );
      });
    });

    describe('intersectionObserver', () => {
      const mockIntersectionObserver = jest.fn();
      const observe = jest.fn();
      const disconnect = jest.fn();
      const onIntersect = jest.fn();

      mockIntersectionObserver.mockReturnValue({
        observe: observe,
        disconnect: disconnect
      });

      window.IntersectionObserver = mockIntersectionObserver;

      test('should not call onIntersect method if not intersected', () => {
        act(() => {
          testRenderer.update(
            <Footer
              areMoviesLoaded={false}
              currentPage={2}
              totalPages={5}
              onIntersect={onIntersect}
            />
          );
        });

        const mockEntry = { isIntersecting: false }
        const observerCallback = mockIntersectionObserver.mock.calls[0][0];
        observerCallback([mockEntry])
        expect(onIntersect).not.toHaveBeenCalled();
      });

      test('should call onIntersect method if is intersected', () => {
        act(() => {
          testRenderer.update(
            <Footer
              areMoviesLoaded={true}
              currentPage={2}
              totalPages={5}
              onIntersect={onIntersect}
            />
          );
        });

        const mockEntry = { isIntersecting: true }
        const observerCallback = mockIntersectionObserver.mock.calls[0][0];
        observerCallback([mockEntry])
        expect(onIntersect).toHaveBeenCalled();
        expect(disconnect).toHaveBeenCalled();
      });
    });
  });
});

