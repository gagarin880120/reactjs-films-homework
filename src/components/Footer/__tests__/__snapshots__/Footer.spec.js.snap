// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Footer component should render without crashing 1`] = `
<div
  className="wrapper"
>
  <div
    className="header"
  >
    films
  </div>
  <p
    className="text"
  >
    Copyright Â© 2020 FILMS. Andrei Mandryk
  </p>
</div>
`;
